core/metadata_vec.cpp

```diff
@@ -0,0 +1,265 @@
+
+#include <algorithm>
+#include "metadata_vec.h"
+
+MetaDataVec::MetaDataVec() {
+    init(std::vector<MDLabel>());
+}
+
+MetaDataVec::MetaDataVec(const std::vector<MDLabel> &labelsVector) {
+    init(labelsVector);
+}
+
+MetaDataVec::MetaDataVec(const FileName &fileName, const std::vector<MDLabel> &desiredLabels) {
+    init(desiredLabels);
+    read(fileName);
+}
+
+MetaDataVec::MetaDataVec(const FileName &fileName) {
+    init({});
+    read(fileName);
+}
+
+MetaDataVec::MetaDataVec(const MetaData &md) {
+    *this = md;
+}
+
+MetaDataVec& MetaDataVec::operator=(const MetaData &md) {
+    this->copyInfo(md);
+    for (const auto& row : md)
+        this->addRow(row);
+    return *this;
+}
+
+void MetaDataVec::init(const std::vector<MDLabel> &labelsVector) {
+    size_t col = 0;
+    for (const auto label : labelsVector) {
+        _label_to_col[label] = col;
+        col++;
+    }
+    _no_columns = col;
+}
+
+void MetaDataVec::read(const FileName &inFile, const std::vector<MDLabel> *dediredLables, bool decomposeStack) {
+    throw NotImplemented();
+}
+
+void MetaDataVec::write(const FileName &outFile, WriteModeMetaData mode) const {
+    String blockName;
+    FileName extFile;
+    FileName _outFile;
+
+    blockName = outFile.getBlockName();
+    if (blockName.empty())
+        blockName = DEFAULT_BLOCK_NAME;
+    _outFile = outFile.removeBlockName();
+    extFile = outFile.getExtension();
+
+    if (extFile == "xml") {
+        writeXML(_outFile, blockName, mode);
+    } else if (extFile == "sqlite") {
+        throw NotImplemented();
+    } else {
+        writeStar(_outFile, blockName, mode);
+    }
+}
+
+void MetaDataVec::clear() {
+    this->baseClear();
+    this->_rows.clear();
+    std::fill(this->_label_to_col.begin(), this->_label_to_col.end(), -1);
+    this->_col_to_label.clear();
+    this->_no_columns = 0;
+}
+
+template <typename T>
+size_t MetaDataVec::addRow(const T &row) {
+    for (size_t labeli = 0; labeli < MDL_LAST_LABEL; ++labeli) {
+        MDLabel label = static_cast<MDLabel>(labeli);
+        if (row.containsLabel(label) && !this->containsLabel(label))
+            this->addLabel(label);
+    }
+
+    MetaDataVecRow newRow;
+
+    for (size_t coli = 0; coli < this->_no_columns; coli++)
+        newRow.push_back({this->_col_to_label[coli]});
+
+    for (size_t labeli = 0; labeli < MDL_LAST_LABEL; ++labeli) {
+        MDLabel label = static_cast<MDLabel>(labeli);
+        if (row.containsLabel(label)) {
+            size_t ourCol = _label_to_col[label];
+            row.getValue(newRow[ourCol]);
+        }
+    }
+
+    _rows.push_back(newRow);
+    return getRowId(_rows.size()-1); // FIXME: should id be changed or kept same?
+}
+
+void MetaDataVec::addRows(const std::vector<MDRowVec> &rows) {
+    for (const auto row : rows)
+        this->addRow(row);
+}
+
+int MetaData::getMaxStringLength(const MDLabel thisLabel) const {
+    return 0xFF;
```

Comment:
I would prefer decimal number

Reply:
Tss, you should be glad i didn't write it in octal! :D
Seriously, I wrote `0xFF` because it explains *why* i chose this number.

Reply:
Why did you chose 255 then? For int, it's not any special value ...

Reply:
Because of space for `\0` in string buffers.

Reply:
Rewritten to decimal.